src/main/drivers/rx_nrf24l01.c

```diff
@@ -0,0 +1,301 @@
+{
+    ENABLE_NRF24();
+    const uint8_t ret = spiTransferByte(NRF24_SPI_INSTANCE, W_REGISTER | ( REGISTER_MASK & reg));
+    for (uint8_t i = 0; i < length; i++) {
+        spiTransferByte(NRF24_SPI_INSTANCE, data[i]);
+    }
+    DISABLE_NRF24();
+    return ret;
+}
+
+/*
+ * Transfer the payload to the nRF24L01 TX FIFO
+ * Packets in the TX FIFO are transmitted when the
+ * nRF24L01 next enters TX mode
+ */
+uint8_t NRF24L01_WritePayload(const uint8_t *data, uint8_t length)
+{
+    ENABLE_NRF24();
+    const uint8_t ret = spiTransferByte(NRF24_SPI_INSTANCE, W_TX_PAYLOAD);
+    for (uint8_t i = 0; i < length; i++) {
+        spiTransferByte(NRF24_SPI_INSTANCE, data[i]);
+    }
+    DISABLE_NRF24();
+    return ret;
+}
+
+uint8_t NRF24L01_ReadReg(uint8_t reg)
+{
+    ENABLE_NRF24();
+    spiTransferByte(NRF24_SPI_INSTANCE, R_REGISTER | (REGISTER_MASK & reg));
+    const uint8_t ret = spiTransferByte(NRF24_SPI_INSTANCE, NOP);
+    DISABLE_NRF24();
+    return ret;
+}
+
+uint8_t NRF24L01_ReadRegisterMulti(uint8_t reg, uint8_t *data, uint8_t length)
+{
+    ENABLE_NRF24();
+    const uint8_t ret = spiTransferByte(NRF24_SPI_INSTANCE, R_REGISTER | (REGISTER_MASK & reg));
+    for (uint8_t i = 0; i < length; i++) {
+        data[i] = spiTransferByte(NRF24_SPI_INSTANCE, NOP);
+    }
+    DISABLE_NRF24();
+    return ret;
+}
+
+/*
+ * Read a packet from the nRF24L01 RX FIFO.
+ */
+uint8_t NRF24L01_ReadPayload(uint8_t *data, uint8_t length)
+{
+    ENABLE_NRF24();
+    const uint8_t ret = spiTransferByte(NRF24_SPI_INSTANCE, R_RX_PAYLOAD);
+    for (uint8_t i = 0; i < length; i++) {
+        data[i] = spiTransferByte(NRF24_SPI_INSTANCE, NOP);
+    }
+    DISABLE_NRF24();
+    return ret;
+}
+
+/*
+ * Empty the transmit FIFO buffer.
+ */
+void NRF24L01_FlushTx()
+{
+    ENABLE_NRF24();
+    spiTransferByte(NRF24_SPI_INSTANCE, FLUSH_TX);
+    DISABLE_NRF24();
+}
+
+/*
+ * Empty the receive FIFO buffer.
+ */
+void NRF24L01_FlushRx()
+{
+    ENABLE_NRF24();
+    spiTransferByte(NRF24_SPI_INSTANCE, FLUSH_RX);
+    DISABLE_NRF24();
+}
+
+#endif // UNIT_TEST
+
+/*
+ * Enter standby mode
+ */
+void NRF24L01_SetStandbyMode(void)
+{
+    // set CE low and clear the PRIM_RX bit to enter standby mode
+    NRF24_CE_LO();
+    NRF24L01_WriteReg(NRF24L01_00_CONFIG, standbyConfig);
+}
+
+/*
+ * Enter receive mode
+ */
+void NRF24L01_SetRxMode(void)
+{
+    NRF24_CE_LO(); // drop into standby mode
+    // set the PRIM_RX bit
+    NRF24L01_WriteReg(NRF24L01_00_CONFIG, standbyConfig | BV(NRF24L01_00_CONFIG_PRIM_RX));
+    NRF24L01_ClearAllInterrupts();
+    // finally set CE high to start enter RX mode
+    NRF24_CE_HI();
+    // nRF24L01+ will now transition from Standby mode to RX mode after 130 microseconds settling time
+}
+
+/*
+ * Enter transmit mode. Anything in the transmit FIFO will be transmitted.
+ */
+void NRF24L01_SetTxMode(void)
+{
+    // Ensure in standby mode, since can only enter TX mode from standby mode
+    NRF24L01_SetStandbyMode();
```

Comment:
This will probably not leave TX mode (CE is low, NRF24L01_00_CONFIG has standby value, so SetStandbyMode is effectively NOP). But it will leave RX, so it should stay in the code.

Pulsing CE during TX has probably no effect

Caller must ensure that last TX operation has ended before transmitting next packet